src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+        this(value, List.of(observers));
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     */
+    public DeferredProperty(T value) {
+        this(value, new ArrayList<>());
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>} with an initial value of <code>null</code> and no observers.
+     */
+    public DeferredProperty() {
+        this(null, new ArrayList<>());
+    }
+
+    /**
+     * Bypasses the dirty checks and just returns the value of {@link DeferredProperty#value}.
+     *
+     * @return The current value of {@link DeferredProperty#value}.
+     */
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Marks this instance as dirty and only if this instance was initially marked as dirty do we propagate this to its {@link DeferredProperty#observers}.
+     *
+     * @return <code>True</code> if we marked this instance as dirty.
+     */
+    public boolean markAsDirty() {
+        // Only when this instance is clean do we want to make it dirty.
+        if (!isDirty()) {
+            isDirty = true;
+
+            // Mark the observers as dirty as they inherit it from this instance.
+            // As it is impossible to add observers after construction then a livelock would never happen.
+            for (DeferredProperty<?> property : observers) {
+                property.markAsDirty();
+            }
+
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Sets the dirty flag as <code>false</code>. If this instance was dirty then true si removed to indicate the dirty flag change from <code>true</code> to <code>false</code>.
+     * Unlike {@link DeferredProperty#markAsDirty()} this does not propagate the cleaning to its observers. For this reason, observers that are dirty will not be cleaned.
+     *
+     * @return <code>True</code> if the dirty flag changed to <code>true</code> from <code>false</code>.
+     */
+    public boolean clean() {
+        if (isDirty()) {
+            isDirty = false;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * If this instance {@link DeferredProperty#isDirty} then we run the {@link Runnable} and then {@link DeferredProperty#clean()}
+     * This should be useed if the dirty state of this instance can be controled by performing an operation rather than setting a new {@link DeferredProperty#value}.
+     *
+     * @param runnable The runnable which is run if we cleaned the object.
+     * @return <code>True</code> if the dirty flag changed to <code>true</code> from <code>false</code>.
+     */
+    public boolean clean(Runnable runnable) {
+        boolean cleaned = clean();
+        if (cleaned) {
+            runnable.run();
+        }
+        return cleaned;
+    }
+
+    /**
+     * Used to check if this instance is marked as dirty.
```

Comment:
```suggestion
     * Checks if this instance is marked as dirty.
```

This is the first time I notice that I prefer that function descriptions start with a verb (because functions do something). Since we don't have a code style yet (see #76), it's up to you whether you want to change all first lines of function descriptions.

Reply:
That is a very good point. I will change all function descriptions in ``DeferredProperty`` to start with a verb and ensure that all fields starts with a noun.